src/pages/index.js

```diff
@@ -1,68 +1,204 @@
+
+// Instances
 const handleImageClick = (link, alt, name) => {
   imagePopup.open({ link, alt, name });
 };
 
+const handleDeleteClick = (cardInstance, cardId) => {
+  const confirmDeleteModal = document.querySelector("#modal-confirm-delete");
+  const confirmDeleteButton = confirmDeleteModal.querySelector("#confirm-delete-button");
+
+  confirmDeleteModal.classList.add("modal_opened");
+
+  confirmDeleteButton.onclick = () => {
+    api
+      .deleteCard(cardId)
+      .then(() => {
+        cardInstance.removeCard();
+        closeModal(confirmDeleteModal);
+      })
+      .catch((err) => console.error(err));
+  };
+
+  const closeButton = confirmDeleteModal.querySelector(".modal__close-button");
+  closeButton.onclick = () => closeModal(confirmDeleteModal);
+};
+
+const handleLikeClick = (cardId, isLiked) => {
+  if (isLiked) {
+    return api.likeCard(cardId).then((data) => {
+      return { ...data, isLiked: true };
+    });
+  } else {
+    return api.dislikeCard(cardId).then((data) => {
+      return { ...data, isLiked: false };
+    });
+  }
+};
+
 const userInfo = new UserInfo({
   nameSelector: ".profile__name",
   jobSelector: ".profile__description",
+  avatarSelector: "#profile-image",
 });
 
 const renderCard = (item) => {
-  const card = new Card(item, "#card-template", handleImageClick);
+  const card = new Card(
+    {
+      name: item.name,
+      link: item.link,
+      alt: item.alt,
+      _id: item._id,
+      likes: item.likes,
+      isLiked: item.isLiked,
+    },
+    "#card-template",
+    handleImageClick,
+    handleDeleteClick,
+    handleLikeClick,
+    userInfo
+  );
   const cardElement = card.getView();
   cardList.addItem(cardElement);
 };
 
 const cardList = new Section(
   {
-    items: initialCards.reverse(),
+    items: [],
     renderer: renderCard,
   },
   ".cards__list"
 );
 
-cardList.renderItems();
+// Extend PopupWithForm for loading state handling
+class CustomPopupWithForm extends PopupWithForm {
+  setLoadingState(isLoading) {
+    const saveButton = this._popup.querySelector(".modal__button");
+
+    if (isLoading) {
+      saveButton.textContent = "Saving...";
+      saveButton.disabled = true;
+    } else {
+      saveButton.textContent = "Save";
+      saveButton.disabled = false;
+    }
+  }
+}
+
+const addCardPopup = new CustomPopupWithForm("#modal-add-card", (formData) => {
+  addCardPopup.setLoadingState(true);
 
-const addCardPopup = new PopupWithForm("#modal-add-card", (formData) => {
-  renderCard({ name: formData.title, link: formData.url, alt: formData.title });
-  addCardPopup.close();
+  api
+    .addCard({ name: formData.title, link: formData.url })
+    .then((data) => {
+      renderCard(data);
+      addCardPopup.close();
+    })
+    .catch((err) => console.error(err))
+    .finally(() => {
+      addCardPopup.setLoadingState(false);
+    });
+});
+
+const editPfpPopup = new CustomPopupWithForm("#modal-edit-pfp", (formData) => {
+  editPfpPopup.setLoadingState(true);
```

Comment:
**Could be improved**

If it’s interesting for you here is how we can make a universal function for handling any submit. We can get rid of such duplicating as **loading effect**, **closing** and **catching errors**


```js

// You can make a universal function that accepts a request, popup instance and optional loading text  
function handleSubmit(request, popupInstance, loadingText = "Saving...") {

  // here we change the button text
  popupInstance.renderLoading(true, loadingText);
  request()
    .then(() => {
      // We need to close only in `then`
      popupInstance.close()
    })
      // we need to catch possible errors
      // console.error is used to handle errors if you don’t have any other ways for that
     .catch(console.error)
    // in `finally` we need to return the initial button text back in any case
    .finally(() => {
      popupInstance.renderLoading(false);
    });
}



```


You need to pass the **request** into the call of `handleSubmit` like that:



```js

// here is an example of the profile form handling
function handleProfileFormSubmit(inputValues) {
  // we create a function that returns a promise
  function makeRequest() {
    // `return` lets us use a promise chain `then, catch, finally` inside `handleSubmit`
    return api.editProfile(inputValues).then((userData) => {
      userInfo.setUserInfo(userData)
    });
  }
  // Here we call the function passing the request, popup instance and if we need some other loading text we can pass it as the 3rd argument
  handleSubmit(makeRequest, profilePopup);
}



```